pkg/credentials/filter/ring_buffer.go

```diff
@@ -0,0 +1,88 @@
+package filter
+
+import (
+	"math"
+)
+
+// RingBuffer implements a simple ring buffer with a static maximal length.
+type RingBuffer struct {
+	start         int
+	end           int
+	currentLength int
+	buffer        []byte
+}
+
+// NewRingBuffer creates a RingBuffer with the given length.
+func NewRingBuffer(length int) *RingBuffer {
+	return &RingBuffer{
+		start:         0,
+		end:           -1,
+		currentLength: 0,
+		buffer:        make([]byte, length),
+	}
+}
+
+// IsFull checks if the buffer is full.
+func (buffer *RingBuffer) IsFull() bool {
+	return buffer.currentLength == len(buffer.buffer)
+}
+
+// IsEmpty checks if the buffer is empty.
+func (buffer *RingBuffer) IsEmpty() bool {
+	return buffer.currentLength == 0
+}
+
+// PopFirst removes the first byte from the buffer and returns it.
```

Comment:
What happens if you try to pop from an empty buffer? can you add this info to the docstring, and add tests for this?

Reply:
I added proper error handling.